File: day20.py
#!/usr/bin/env python3

day20_input = 34000000


def number_of_presents_part1(house):
    return sum(elf * 10 for elf in range(1, house + 1) if house % elf == 0)


def number_of_presents_part2(house):
    return sum(elf * 11 for elf in range(1, house + 1) if house % elf == 0 and house // elf <= 50)


def part1():
    house = 1
    high = -1

    while high < day20_input:
        presents = number_of_presents_part1(house)

        if presents > high:
            high = presents
            print("New high: {} at house: {}".format(high, house))

        if presents < day20_input:
            house += 1

    print("House: {}".format(house))


def part2():
    house = 1
    high = -1

    while high < day20_input:
        presents = number_of_presents_part2(house)

        if presents > high:
            high = presents
            print("New high: {} at house: {}".format(high, house))

        if presents < day20_input:
            house += 1

    print("House: {}".format(house))


part1()
part2()

File: day20.cpp
#include <iostream>

const int day20_input = 34000000;

int number_of_presents_part1(int house)
{
    int presents = 0;
    for (int elf = 1; elf <= house; elf++) {
        if (house % elf == 0) {
            presents += elf * 10;
        }
    }
    return presents;
}

int number_of_presents_part2(int house)
{
    int presents = 0;
    for (int elf = 1; elf <= house; elf++) {
        if (house % elf == 0 && house / elf <= 50) {
            presents += elf * 11;
        }
    }
    return presents;
}

void part1()
{
    int house = 0;
    int high = 0;

    while (high < day20_input) {
        int presents = number_of_presents_part1(house);

        if (presents > high) {
            high = presents;
            std::cout << "New high: " << high << " at house: " << house << std::endl;
        }

        if (presents < day20_input) {
            house++;
        }
    }

    std::cout << "House: " << house << std::endl;
}

void part2()
{
    int house = 0;
    int high = 0;

    while (high < day20_input) {
        int presents = number_of_presents_part2(house);

        if (presents > high) {
            high = presents;
            std::cout << "New high: " << high << " at house: " << house << std::endl;
        }

        if (presents < day20_input) {
            house++;
        }
    }

    std::cout << "House: " << house << std::endl;
}

int main()
{
    part1();
    part2();

    return 0;
}
